File: apps/demo-be-python-fastapi/tests/integration/steps/security_steps.py
"""BDD step definitions for security feature."""

from fastapi.testclient import TestClient
from pytest_bdd import given, parsers, scenarios, then, when

from demo_be_python_fastapi.config import settings
from tests.integration.conftest import GHERKIN_ROOT

scenarios(str(GHERKIN_ROOT / "security" / "security.feature"))

_PASSWORD = "Str0ng#Pass1"
_ADMIN_PASSWORD = "Admin#Str0ng1"


def _promote_to_admin(client: TestClient, username: str) -> None:
    """Promote a user to ADMIN by using the internal SQLAlchemy DB directly.

    Since we don't have a separate admin-creation endpoint, we use a special
    registration endpoint pattern or directly mutate via the test DB.
    Instead, we register via a special fixture that creates an ADMIN user directly.
    """
    pass  # handled via _register_admin_user below


def _register_and_promote_admin(client: TestClient, username: str, password: str) -> dict:
    """Register a user and immediately set their role to ADMIN.

    This uses the app's internal DI to get the DB session and update the role.
    We achieve this by accessing the app's dependency override to get a session.
    """
    # Register as normal user first
    resp = client.post(
        "/api/v1/auth/register",
        json={"username": username, "email": f"{username}@example.com", "password": password},
    )
    assert resp.status_code == 201, f"Registration failed: {resp.text}"
    user_data = resp.json()

    # Directly update role using the app's session factory via the test client's app
    from sqlalchemy.orm import Session

    app = client.app  # type: ignore[attr-defined]
    override = app.dependency_overrides.get(  # type: ignore[union-attr]
        __import__("demo_be_python_fastapi.dependencies", fromlist=["get_db"]).get_db
    )
    if override:
        db_gen = override()
        db: Session = next(db_gen)
        try:
            from demo_be_python_fastapi.infrastructure.models import UserModel

            user = db.get(UserModel, user_data["id"])
            if user:
                user.role = "ADMIN"
                db.commit()
        finally:
            try:
                next(db_gen)
            except StopIteration:
                pass

    return user_data


@given(
    parsers.parse('a user "{username}" is registered and locked after too many failed logins'),
    target_fixture="locked_user",
)
def register_and_lock_user(client: TestClient, username: str) -> dict:
    # Register
    resp = client.post(
        "/api/v1/auth/register",
        json={
            "username": username,
            "email": f"{username}@example.com",
            "password": _PASSWORD,
        },
    )
    assert resp.status_code == 201
    user_data = resp.json()

    # Fail logins to lock the account
    for _ in range(settings.max_failed_login_attempts):
        client.post(
            "/api/v1/auth/login",
            json={"username": username, "password": "WrongPass#1234"},
        )
    return user_data


@given(
    parsers.parse('an admin user "{username}" is registered and logged in'),
    target_fixture="admin_tokens",
)
def register_admin_and_login(client: TestClient, username: str) -> dict:
    user_data = _register_and_promote_admin(client, username, _ADMIN_PASSWORD)
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": username, "password": _ADMIN_PASSWORD},
    )
    assert resp.status_code == 200, f"Admin login failed: {resp.text}"
    return {**resp.json(), "id": user_data["id"]}


@given(
    parsers.parse('"alice" has had the maximum number of failed login attempts'),
)
def alice_max_failed_attempts(client: TestClient, registered_user: dict) -> None:
    for _ in range(settings.max_failed_login_attempts):
        client.post(
            "/api/v1/auth/login",
            json={"username": "alice", "password": "WrongPass#1234"},
        )


@given("an admin has unlocked alice's account")
def admin_unlocks_alice(client: TestClient, locked_user: dict) -> None:
    # Create a temporary admin inline for this step
    _register_and_promote_admin(client, "tmpadmin", _ADMIN_PASSWORD)
    login_resp = client.post(
        "/api/v1/auth/login",
        json={"username": "tmpadmin", "password": _ADMIN_PASSWORD},
    )
    assert login_resp.status_code == 200, f"Admin login failed: {login_resp.text}"
    tmp_tokens = login_resp.json()
    resp = client.post(
        f"/api/v1/admin/users/{locked_user['id']}/unlock",
        headers={"Authorization": f"Bearer {tmp_tokens['access_token']}"},
    )
    assert resp.status_code == 200, f"Unlock failed: {resp.text}"


# --- When steps ---


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/unlock"),
    target_fixture="response",
)
def admin_unlock_alice(client: TestClient, locked_user: dict, admin_tokens: dict):  # type: ignore[no-untyped-def]
    return client.post(
        f"/api/v1/admin/users/{locked_user['id']}/unlock",
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )


# --- Then steps ---


@then('alice\'s account status should be "locked"')
def check_alice_locked(client: TestClient, registered_user: dict) -> None:
    # Verify the account is locked by attempting login — locked accounts return 401
    # The previous When step already confirmed the 401; attempt another login to verify lock
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": "alice", "password": _PASSWORD},
    )
    assert resp.status_code == 401, (
        f"Expected 401 for locked account, got {resp.status_code}: {resp.text}"
    )

File: apps/crud-be-rust-axum/tests/integration/steps/security_steps.rs
use cucumber::{given, when};

use crate::world::AppWorld;

#[when(
    regex = r#"the client sends POST /api/v1/auth/register with body \{ "username": "alice", "email": "alice@example\.com", "password": "Short1!Ab" \}"#
)]
async fn register_short_password(world: &mut AppWorld) {
    world
        .svc_register("alice", "alice@example.com", "Short1!Ab")
        .await;
}

#[when(
    regex = r#"the client sends POST /api/v1/auth/register with body \{ "username": "alice", "email": "alice@example\.com", "password": "AllUpperCase1234" \}"#
)]
async fn register_no_special_char(world: &mut AppWorld) {
    world
        .svc_register("alice", "alice@example.com", "AllUpperCase1234")
        .await;
}

#[given(expr = "{string} has had the maximum number of failed login attempts")]
async fn max_failed_attempts(world: &mut AppWorld, username: String) {
    for _ in 0..5 {
        world.svc_login(&username, "WrongPass!123").await;
    }
}

#[given(expr = "a user {string} is registered and locked after too many failed logins")]
async fn register_and_lock(world: &mut AppWorld, username: String) {
    let email = format!("{username}@example.com");
    world.svc_register(&username, &email, "Str0ng#Pass1").await;
    if world.last_status == 201 && username == "alice" {
        world.alice_id = world
            .last_body
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| uuid::Uuid::parse_str(s).ok());
    }

    for _ in 0..5 {
        world.svc_login(&username, "WrongPass!123").await;
    }
}

#[given(expr = "an admin user {string} is registered and logged in")]
async fn register_admin_and_login(world: &mut AppWorld, username: String) {
    let email = format!("{username}@example.com");
    world.svc_register(&username, &email, "Str0ng#Pass1").await;

    if world.last_status == 201 {
        let admin_id = world
            .last_body
            .get("id")
            .and_then(|v| v.as_str())
            .and_then(|s| uuid::Uuid::parse_str(s).ok());

        if let Some(admin_uuid) = admin_id {
            world.promote_to_admin(admin_uuid).await.unwrap();
        }

        world.svc_login(&username, "Str0ng#Pass1").await;
        if world.last_status == 200 {
            world.admin_token = world
                .last_body
                .get("accessToken")
                .and_then(|v| v.as_str())
                .map(String::from);
        }
    }
}

#[given("an admin has unlocked alice's account")]
async fn admin_unlocked_alice(world: &mut AppWorld) {
    // Register a sysadmin if we don't have an admin token yet
    if world.admin_token.is_none() {
        world
            .svc_register("sysadmin", "sysadmin@example.com", "Str0ng#Pass1")
            .await;
        if world.last_status == 201 {
            let admin_id = world
                .last_body
                .get("id")
                .and_then(|v| v.as_str())
                .and_then(|s| uuid::Uuid::parse_str(s).ok());
            if let Some(admin_uuid) = admin_id {
                world.promote_to_admin(admin_uuid).await.unwrap();
            }
            world.svc_login("sysadmin", "Str0ng#Pass1").await;
            if world.last_status == 200 {
                world.admin_token = world
                    .last_body
                    .get("accessToken")
                    .and_then(|v| v.as_str())
                    .map(String::from);
            }
        }
    }

    let admin_bearer = world.admin_bearer();
    let alice_id = match world.alice_id {
        Some(id) => id,
        None => return,
    };
    if !admin_bearer.is_empty() {
        world.svc_admin_unlock_user(&admin_bearer, alice_id).await;
    }
}

#[when("the admin sends POST /api/v1/admin/users/{alice_id}/unlock")]
async fn admin_unlock_alice(world: &mut AppWorld) {
    let admin_bearer = world.admin_bearer();
    let alice_id = match world.alice_id {
        Some(id) => id,
        None => return,
    };
    world.svc_admin_unlock_user(&admin_bearer, alice_id).await;
}
